E2E open-sourced

rnugmanx opened the pull request.

rkazants commented on tests/e2e_oss/.automation/env_config.yml:
```diff
@@ -0,0 +1,43 @@
+# PyTorch specific environment
+pytorch_internal_models:  ${SHARE}/models/internal/pytorch
+pytorch_pretrained_models_path: ${SHARE}/models/internal/pytorch/pretrained/0.7.4
+pytorch_torchvision_models_path: ${SHARE}/models/internal/pytorch/torchvision/0.2.1
+pytorch_timm_models_path: ${SHARE}/models/internal/pytorch/timm
+pytorch_hf_models_path:  ${SHARE}/models/internal/pytorch/huggingface
+pytorch_to_onnx_dump_path: ${SHARE}/models/internal/pytorch/pytorch_to_onnx_dumped
+
+# private models location
+private_models: ${SHARE}/models/private/
+
+# large models location
+tf_large_models: ${SHARE_BIG_MODELS}/models/tf/
+pytorch_large_models: ${SHARE_BIG_MODELS}/models/pytorch/
+
+# ICV models zoo location
+icv_model_zoo_models: ${SHARE}/models/icv_modelzoo/
```
Not sure that all of these config files are needed at all

rkazants reviewed (changes requested):
Disagree with these tests because our agreement was to have convenient API for adding new e2e tests.
Here you just move private e2e repository code.

Let us do as follows: create README how to add e2e test for new model. Imagine some external contributor wants to do this.
Based on this documentation, we will see what we should do with test API and how easy it is.

Also, we need to do smth with multiple config files. Think about platforms that have been already deprecated so we can remove it from here.

I still think that we need do the test system almost from scratch or independent from existing one in the private repository. That is because the legacy system drag a lot of legacy stuff (platforms, APIs) that is over-complicating the system.

rkazants commented on tests/e2e_oss/pipelines/pipeline_base_classes/caffe_eltwise.py:
```diff
@@ -0,0 +1,39 @@
+from tests.e2e_oss.pipelines.pipeline_templates.ir_gen_templates import common_ir_generation
+from tests.e2e_oss.pipelines.pipeline_templates.preproc_templates import assemble_preproc_caffe
+from tests.e2e_oss._utils.path_utils import ref_from_model, prepend_with_env_path, resolve_file_path
+
+common_input_file = resolve_file_path("test_data/inputs/caffe/object_detection_voc.npz")
+
+
```
Why do we need Caffe and legacy models here at all?

rkazants commented on tests/e2e_oss/pipelines_custom/LSTM_no_seqlen/mxnet/synthetic_lstm.py:
```diff
@@ -0,0 +1,81 @@
+from collections import OrderedDict
```
MXNet?
We don't need it.

rkazants commented on tests/utils/e2e/ref_collector/score_mxnet.py:
```diff
@@ -0,0 +1,101 @@
+os.environ['GLOG_minloglevel'] = '3'
+
+
+class ScoreMxnet(ClassProvider):
```
MXNet. please remove it

rkazants commented on tests/e2e_oss/utils/kaldi_utils.py:
```diff
@@ -0,0 +1,47 @@
+    from kaldiio import load_ark
+except ImportError:
+    pass
+
```
kaldi. Please remove it

rkazants commented on tests/utils/e2e/comparator/eltwise_kaldi.py:
```diff
@@ -0,0 +1,78 @@
+from .provider import ClassProvider
+
+
+class EltwiseKaldiComparator(ClassProvider):
```
Kaldi, please remove it. I think eltwise comparation should be generic for all FWs

rkazants commented on tests/utils/e2e/postprocessors/speech_kaldi.py:
```diff
@@ -0,0 +1,33 @@
+    from kaldiio import save_ark
+except ImportError:
+    pass
+
```
Not needed, please remove it

rkazants commented on tests/utils/e2e/ref_collector/score_caffe.py:
```diff
@@ -0,0 +1,61 @@
+from .provider import ClassProvider
+
+os.environ['GLOG_minloglevel'] = '3'
+
```
caffe, not needed

rkazants commented on tests/e2e_oss/requirements.txt:
```diff
@@ -30,6 +30,7 @@ pytest-xdist==2.1.0
 pytest-timeout==1.3.2
 
 # for common utils, e2e_oss
+openvino-dev
```
if possible, please clean-up test core part from legacy stuff. And let us try to test `openvino` here

rkazants reviewed (changes requested):
The directory structure is not clear. Let us simplify in this PR.
Just create a directory:
- e2e
    |-----utils
    |-----common
    |-----tensorflow
    |-----pytorch
    |-----run_tests.py # common entry point

Also, let us have just one `README.md` (not several) and clearly explain how to add new model for testing. And what options `run_tests.py` has.

rnugmanx commented:
Removed  readme in pipelines folder, remaining readme in help_scripts and readme in the root folder have different customers so I left it as is.
Also I added info about how to add test for model from tf hub repo. For possible options there is 'pytest test.py -h' option which is already described in readme. Removed some redundant info from readme regarding advanced entrypoints

Structure of e2e tests now:
e2e
|-----test_utils
|-----common
|-----pipelines
    |-----tensorflow
    |-----pytorch (will be added)
|-----run_tests.py # common entry point

rkazants commented on .github/workflows/linux.yml:
```diff
@@ -481,6 +481,79 @@ jobs:
       runner: ${{ github.event_name == 'schedule' && 'ubuntu-20.04-16-cores' || 'ubuntu-20.04-8-cores' }}
```
I still see this structure:
![image](https://github.com/openvinotoolkit/openvino/assets/35459624/0324b9d9-9c79-410f-b32b-b3b2c3847efb)

rnugmanx replied:
I moved it under common section

rkazants commented on tests/e2e_oss/test_utils/kaldi_utils.py:
```diff
@@ -0,0 +1,47 @@
+    from kaldiio import load_ark
+except ImportError:
+    pass
+
```
please remove

rnugmanx replied:
Done

rnugmanx commented on tests/CMakeLists.txt:
```diff
@@ -5,3 +5,5 @@
 add_subdirectory(layer_tests)
 add_subdirectory(model_hub_tests)
 add_subdirectory(samples_tests)
+add_subdirectory(e2e_oss)
```
Line removed, renaming will be done in next PR

rkazants replied:
why is it difficult to rename it? Let us do it now because anyway closed source does not depend on it.

rnugmanx commented on tests/utils/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Copyright (C) 2018-2023 Intel Corporation
```
This file is responsible for building utils folder in precommit

rkazants commented on .github/workflows/linux.yml:
```diff
@@ -499,6 +499,79 @@ jobs:
       runner: ${{ github.event_name == 'schedule' && 'ubuntu-20.04-16-cores' || 'ubuntu-20.04-8-cores' }}
       event: ${{ github.event_name }}
 
+#  E2E_TensorFlow_Models_Tests:
+#    name: E2E TensorFlow Hub Models tests
+#    needs: Build
+#    defaults:
+#      run:
+#        shell: bash
+#    runs-on: ${{ github.event_name == 'schedule' && 'ubuntu-20.04-16-cores' || 'ubuntu-20.04-8-cores'}}
+#    env:
+#      INSTALL_DIR: ${{ github.workspace }}/install
+#      INSTALL_TEST_DIR: ${{ github.workspace }}/install/tests
```
let us remove it for now and add when you will activate it in GHA. Now I afraid that this is not working

rnugmanx replied:
Done

rkazants commented on tests/e2e_oss/README.md:
```diff
@@ -0,0 +1,65 @@
+# E2E OSS tests
```
```suggestion
# End-to-end Tests User Documentation
```

rkazants commented on tests/e2e_oss/README.md:
```diff
@@ -0,0 +1,65 @@
+# E2E OSS tests
+
+This folder contains E2E test framework code, E2E tests and related materials
+(reference outputs, test data, etc.)
```
```suggestion
This folder contains a code to run end-to-end validation of OpenVINO on real models of different frameworks (PyTorch, TensorFlow, and ONNX). Validation steps include conversion of a real model to `ov::Model`, model inference on data samples using OpenVINO, and comparison inference results with a reference results from original framework.

The documentation provides necessary information about environment setup for e2e validation run, adding new model to the validation, and instructions to launch validation. 
```

rkazants commented on tests/e2e_oss/README.md:
```diff
@@ -0,0 +1,65 @@
+
+This folder contains E2E test framework code, E2E tests and related materials
+(reference outputs, test data, etc.)
+
```
Let us have the following sections in README:
1. Environment setup
2. Adding new model
3. Launch validation

rnugmanx replied:
I fixed all previous comments not related to readme. I suggest to move next changes of readme in next PR when we will make e2e work in precommit, otherwise I think we could discuss about it forever and wont merge this PR

rkazants commented on tests/e2e_oss/README.md:
```diff
@@ -0,0 +1,65 @@
+This folder contains E2E test framework code, E2E tests and related materials
+(reference outputs, test data, etc.)
+
+## Basic scripts
+
+There is one basic script which is responsible for test run with default pipeline - test_base.py. This script perform the
+following actions:
+1. Loads model from its source 
+2. Collects reference from framework
+3. Converts original model through OVC convert model
+4. Performs inference on converted model
+5. Compares reference and output from converted model through element-wise comparison
+
+
+> test_base.py script is run with the help of [pytest](https://docs.pytest.org).
+> The command-line is the same for all the scripts, though some of them
+> __might not__ use specific options.
```
Not clear what are basic scripts. Imagine that some external contributor reads this documentation

rnugmanx replied:
Please refer to my comment above

rkazants commented on tests/e2e_tests/README.md:
````diff
@@ -0,0 +1,50 @@
+To add new test for model from TF Hub repo just add new line into pipelines/production/tf_hub/precommit.yml
+This line should contain: model name, and its link to download. Example:
+```
+my_model_name,https://www.example.com/model
````
```suggestion
movenet/singlepose/lightning,https://www.kaggle.com/models/google/movenet/frameworks/tensorFlow2/variations/singlepose-lightning/versions/4
```

rkazants commented on tests/e2e_tests/README.md:
```diff
@@ -0,0 +1,50 @@
+     
+### Add model from TensorFlow Hub repo to end-to-end validation:
+To add new test for model from TF Hub repo just add new line into pipelines/production/tf_hub/precommit.yml
+This line should contain: model name, and its link to download. Example:
```
```suggestion
This line should contain comma separated model name and its link
```

rkazants commented on tests/e2e_tests/README.md:
```diff
@@ -0,0 +1,50 @@
+
+#### Launch tests
+
+Module [test_base.py](https://github.com/openvinotoolkit/openvino/tree/master/tests/e2e_tests/test_base.py) is main entry-point to run E2E OSS tests.  
```
```suggestion
[test_base.py](https://github.com/openvinotoolkit/openvino/tree/master/tests/e2e_tests/test_base.py) is main script to run end-to-end tests.
```

rkazants commented on tests/e2e_tests/common/preprocessors_tf_hub/preprocessors_tf_hub.py:
```diff
@@ -0,0 +1,50 @@
+from .tf_hub_provider import ClassProvider
+import numpy as np
+import logging as log
+
```
why do we need preprocessor for tf hub?

rnugmanx replied:
thanks, deleted

rkazants commented on tests/e2e_tests/pipelines/tf_hub/tf_hub_case_class.py:
```diff
@@ -0,0 +1,27 @@
+from collections import OrderedDict
```
please put a license:

```
# Copyright (C) 2018-2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

rkazants replied:
we have a lot places with missed license.

rnugmanx replied:
I put it in every .py file
